File: TinyScript/js/src/main/js/parser/SimpleParser.ts
import ASTNode from "./ast/ASTNode";
import PeekTokenIterator from "./PeekTokenIterator";
import Expr from "./ast/Expr";
import ASTNodeTypes from "./ast/ASTNodeTypes";
import Factor from "./ast/Factor";
import ParseException from "./ParseException";

export default class SimpleParser {
    static parse(it: PeekTokenIterator): ASTNode {
        const expr = new Expr(null);
        const scalar = Factor.parse(it);
        if (scalar == null) {
            throw new ParseException("parser exception")
        }
        if (!it.hasNext()) {
            return scalar;
        }
        expr.lexeme = it.peek();
        it.nextMatch("+");
        expr.label = "+";
        expr.addChild(scalar);
        expr.type = ASTNodeTypes.BINARY_EXPR;
        const astNode = SimpleParser.parse(it);
        expr.addChild(astNode);
        return expr;
    }
}

File: TinyScript/java/src/main/java/paser/ast/SimpleParser.java
package paser.ast;

import paser.util.ParseException;
import paser.util.PeekTokenIterator;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/9/3 0:22
 **/
public class SimpleParser {
    public static  ASTNode parse(PeekTokenIterator it) throws ParseException {
        Expr expr = new Expr();
        Scalar scalar = new Scalar(expr,it);
        if (!it.hasNext()){
            return scalar;
        }
        expr.setLexeme(it.peek());
        it.nextMatch("+");
        expr.setLabel("+");
        expr.addChild(scalar);
        expr.setType(ASTNodeTypes.BINARY_EXPR);
        ASTNode rightNode = parse(it);
        expr.addChild(rightNode);
        return expr;
    }
}
